ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProject.java

```diff
@@ -956,33 +951,62 @@ public void refresh() {
     super.repaint();
   }
 
-  // TODO: Accumulate changes instead of sending it every time.
+  private interface TxnSendListener {
+    void onSendCompleted();
+  }
+
+  private final AtomicReference<TxnSendListener> txnSendingListener = new AtomicReference<>();
+
   private Unit sendProjectStateLogs() {
     gpLogger.debug("Sending project state logs");
+    if (txnSendingListener.get() != null) return Unit.INSTANCE;
     try {
       var baseTxnCommitInfo = myBaseTxnCommitInfo.get();
       var txns = myProjectDatabase.fetchTransactions(baseTxnCommitInfo.right + 1, 1);
       if (!txns.isEmpty()) {
-        getWebSocket().sendLogs(new InputXlog(
-          baseTxnCommitInfo.left,
-          "userId",
-          "refid",
-          txns
-        ));
+        var listener = new TxnSendListener() {
+          @Override
+          public void onSendCompleted() {
+            txnSendingListener.compareAndSet(this, null);
+          }
+        };
+        if (txnSendingListener.compareAndSet(null, listener)) {
+          getWebSocket().sendLogs(new InputXlog(
+            baseTxnCommitInfo.left,
+            "userId",
+            "refid",
+            txns
+          ));
+        }
       }
     } catch (ProjectDatabaseException e) {
       gpLogger.error("Failed to send logs", new Object[]{}, ImmutableMap.of(), e);
     }
     return Unit.INSTANCE;
   }
 
+  @Override
+  protected Unit onProjectLogUpdate() {
+    if (isColloboqueLocalTest()) {
+      super.onProjectLogUpdate();
+      sendProjectStateLogs();
+    }
+    return Unit.INSTANCE;
+  }
+
   private Unit fireXlogReceived(ServerCommitResponse response) {
     myBaseTxnCommitInfo.update(response.getBaseTxnId(), response.getNewBaseTxnId(), 1);
+    txnSendingListener.get().onSendCompleted();
+    sendProjectStateLogs();
     return Unit.INSTANCE;
   }
 
   private Unit onBaseTxnIdReceived(String baseTxnId) {
     myBaseTxnCommitInfo.update("", baseTxnId, 0);
+    var listener = txnSendingListener.get();
+    // Websocket is [re-]started. Previous messages are discarded.
+    if (listener != null) listener.onSendCompleted();
+    sendProjectStateLogs();
```

Comment:
It appears that the comment above (previous messages are discarded) and these two lines (sendCompleted, sendLogs) are kinda controversial. We obviously need to make sure that the state which we have locally matches the server state which corresponds to this `baseTxnId`. Until it is not done, we probably want to really discard the changes, that is, clear the state, whatever it is, but not do what we do in the normal success case.

Reply:
Thanks!
Added a TODO for performing synchronization with the server. For now, logs are sent only when the server specifies that the project is empty.